Doc: Add links to iteratee. Fix #2052

Florian-R opened the pull request:
A proposal for #2052. Not sure it's the way to go, but just to get the ball rollin'

Fixes #2052; Resolves https://github.com/jashkenas/underscore/pull/1864

Florian-R commented on index.html:
```diff
@@ -451,10 +451,10 @@ <h2 id="collections">Collection Functions (Arrays or Objects)</h2>
+        Iterates over a <b>list</b> of elements, yielding each in turn to an <a href="#iteratee"><b>iteratee</b></a>
+        function. The <a href="#iteratee"><b>iteratee</b></a> is bound to the <b>context</b> object, if one is
+        passed. Each invocation of <a href="#iteratee"><b>iteratee</b></a> is called with three arguments:
+        <tt>(element, index, list)</tt>. If <b>list</b> is a JavaScript object, <a href="#iteratee"><b>iteratee</b></a>'s
```
Not sure we want to have the link duplicated when we have many occurrence in the same section

captbaritone replied:
I think the Wikipedia approach may work well here. They suggest only adding the link the first time the word occurs. Maybe only make it a link the first time it occurs in the documentation for each function?

captbaritone commented:
Here's a link if would help anyone review: https://rawgit.com/Florian-R/underscore/iteratee-doc/index.html

captbaritone commented:
The code might look cleaner if you re-wrapped the paragraphs.

Florian-R commented:
Thanks for your suggestion @captbaritone. The PR have been updated accordingly. Let me know if you want the commits to be squashed.

jridgewell commented:
@akre54: want to rule here? Seems fine to me, but I don't handle any of the docs.

akre54 commented:
Looks cool to me. Do we need to shore up the difference between `predicate` and `iteratee`?

Florian-R commented:
Added 45607a8 with some copy-editing for test purpose. This commit mainly remove `predicate` in favour of `iteratee`

Florian-R commented on index.html:
```diff
@@ -526,11 +528,11 @@ <h2 id="collections">Collection Functions (Arrays or Objects)</h2>
         <br />
         Looks through each value in the <b>list</b>, returning the first one that
-        passes a truth test (<b>predicate</b>), or <tt>undefined</tt> if no value
+        passes a truth test (<a href="#iteratee"><b>iteratee</b></a>), or <tt>undefined</tt> if no value
```
Maybe simpler to do just here 

```html
satisfy the <a href="#iteratee"><b>iteratee</b></a>, else <tt>undefined</tt>.
```

captbaritone replied:
Let's leave the wording as is for now.

Florian-R commented on index.html:
```diff
@@ -1042,10 +1045,11 @@ <h2 id="arrays">Array Functions</h2>
         <br />
         Similar to <a href="#indexOf"><tt>_.indexOf</tt></a>, returns the first index
-        where the <strong>predicate</strong> truth test passes; otherwise returns <i>-1</i>.
+        of the element passing the truth test (<a href="#iteratee"><b>iteratee</b></a>);
```
Not totally happy with this wording.

megawac commented:
Nice, looks like a better version of my proposed patch a while ago https://github.com/jashkenas/underscore/pull/1864

jdalton commented:
I'm fine with `predicate`. It tells me in just a word that its result is truthy or falsey. A predicate is a kind of iteratee just as an iteratee is a kind of callback.

craigmichaelmartin commented on index.html:
```diff
@@ -479,10 +481,10 @@ <h2 id="collections">Collection Functions (Arrays or Objects)</h2>
-        (or <tt>key</tt>) of the iteration, and finally a reference to the entire
-        <tt>list</tt>.
+        through a transformation function (<a href="#iteratee"><b>iteratee</b></a>).
+        The <tt>iteratee</tt> is passed three arguments: the <tt>value</tt>,
```
nitpick: could you not make the iteratee param an html `<tt>` element? Other than here (and I know it was in the original), the underscore function params are `<b>` elements at first occurrence, and the callback's params are the `<tt>` elements.

Florian-R commented:
Sorry @megawac, I missed #1864 when I opened this PR. This has been updated per your suggestion and @craigmichaelmartin 's one.

@jdalton Did you mean you'd prefer to totally drop 45607a8?

Florian-R commented on index.html:
```diff
@@ -451,11 +451,13 @@ <h2 id="collections">Collection Functions (Arrays or Objects)</h2>
-        Iterates over a <b>list</b> of elements, yielding each in turn to an <b>iteratee</b>
-        function. The <b>iteratee</b> is bound to the <b>context</b> object, if one is
+        Iterates over a <b>list</b> of elements, yielding each in turn to an
+        <a href="#iteratee"><b>iteratee</b></a> function.
```
After a double-check on #1864, I've seen a comment from @megawac stating 

> Each and reduce don't use iteratee - they use createCallback

So this is definitively incorrect. Do we just remove the link or do we adjust the wording here?

megawac replied:
I think the wording needs to be adjusted here

Florian-R replied:
Any good suggestion here? I can't come up with anything better than `callback` instead of `iteratee` . This was previously rejected in #1767 and I'm not sure this deserve much bikeshedding.

captbaritone replied:
Let's just leave this as is for now. Bold __iteratee__ to indicate that it's a variable name, but not linked since it's not transformed through `_.iteratee`. We can revisit this wording in another PR.

jdalton commented:
@Florian-R
> Did you mean you'd prefer to totally drop 45607a8?

Yep. I don't mind links or notes, but I dig how descriptive `predicate` is.

captbaritone commented:
Any progress on this? @Florian-R Are you blocked on anything?

Florian-R commented:
Yap I was waiting for input from @megawac on the adjustments about the wording and somehow totally forgot this PR. If you've got some idea I'll be happy to move forward with this one.

captbaritone commented on index.html:
```diff
@@ -498,7 +500,7 @@ <h2 id="collections">Collection Functions (Arrays or Objects)</h2>
         Also known as <b>inject</b> and <b>foldl</b>, reduce boils down a <b>list</b> of values into a single value.
         <b>Memo</b> is the initial state of the reduction, and each successive step of it should be returned by
-        <b>iteratee</b>. The iteratee is passed four arguments: the <tt>memo</tt>, then the <tt>value</tt> and
+        <a href="#iteratee"><b>iteratee</b></a>. The iteratee is passed four arguments: the <tt>memo</tt>, then the <tt>value</tt> and
```
Similar to `_.each` let's remove this links, but leave it bold.

captbaritone commented on index.html:
```diff
@@ -556,7 +558,7 @@ <h2 id="collections">Collection Functions (Arrays or Objects)</h2>
         <br />
         Looks through each value in the <b>list</b>, returning an array of all
-        the values that contain all of the key-value pairs listed in <b>properties</b>.
+        the values that <a href="#matches">matches</a> the key-value pairs listed in <b>properties</b>.
```
:+1:

captbaritone commented:
@Florian-R Apologies for the continued delays. 

How about we leave the descriptions of all the __predicate__ functions alone and instead add to the end of each one:

> __predicate__ is transformed through __[iteratee](http://underscorejs.org/#iteratee)__ to facilitate shorthand syntaxes.

Florian-R commented:
@captbaritone I like it. There have been several change in the doc recently IIRC so I'll need to take a look at the changes to see if this PR is always 100% accurate. Not quite the time right now but I'll try to tackle this next week.

Florian-R commented:
OK, I have removed  45607a8 (which changed wording from predicate to iteratee) and updated with the suggestion from @captbaritone 

I have mixed feeling on this one, this is cleared but the wording seems very redundant.

@captbaritone thoughts?

Florian-R commented on index.html:
```diff
@@ -456,11 +456,13 @@ <h2 id="collections">Collection Functions (Arrays or Objects)</h2>
-        arguments will be <tt>(value, key, list)</tt>. Returns the <b>list</b> for chaining.
+        <tt>(element, index, list)</tt>. If <b>list</b> is a JavaScript object,
+        <b>iteratee</b>'s arguments will be <tt>(value, key, list)</tt>.
+        Returns the <b>list</b> for chaining.
```
Yap, will do. Also, I've totally forgot about this PR, will try to update it with current master once I'll had some spare time.

Florian-R commented:
@captbaritone Finally took some time to force push some edit. I *think* this is in a good shape, but I've probably missed something.

I've removed some previous re-wrapping when lines weren't that much long. Hopefully it will help to review, and it could be addressed in another PR if needed.

captbaritone commented:
Thanks for sticking with this one! This is a much needed improvement to our docs. If you think there are other places we could make improvements, please feel free to open pull requests.
